cni/pkg/ebpf/app/ambient_redirect.bpf.c

```diff
@@ -0,0 +1,262 @@
+|--------------------------------------------------------------|
+|  0          |     6, 00:11:22:33:44:55                       |
+|--------------------------------------------------------------|
+*/
+
+struct {
+        __uint(type, BPF_MAP_TYPE_ARRAY);
+        __uint(max_entries, 1);
+        __type(key, __u32);
+        __type(value, struct ztunnel_info);
+        __uint(pinning, LIBBPF_PIN_BY_NAME);
+} ztunnel_info SEC(".maps");
+
+/* This is a hash map to store current application info
+|--------------------------------------------------------------------------|
+|  key(pod IP)         |        Val(app info: ifindex, mac address)        |
+|--------------------------------------------------------------------------|
+|  10.101.1.2          |        2, 00:11:22:33:44:aa                       |
+|--------------------------------------------------------------------------|
+|  10.101.2.2          |        3, 00:11:22:33:44:bb                       |
+|--------------------------------------------------------------------------|
+*/
+
+struct {
+        __uint(type, BPF_MAP_TYPE_HASH);
+        __uint(max_entries, APP_INFO_MAP_SIZE);
+        __type(key, __u32);
+        __type(value, struct app_info);
+        __uint(pinning, LIBBPF_PIN_BY_NAME);
+} app_info SEC(".maps");
+
+// ipv4 should be in big endian
+static __inline struct app_info * get_app_info_from_ipv4(__u32 ipv4)
+{
+    return bpf_map_lookup_elem(&app_info, &ipv4);
+}
+
+static __inline struct ztunnel_info * get_ztunnel_info()
+{
+    uint32_t key = 0;
+    return bpf_map_lookup_elem(&ztunnel_info, &key);
+}
+
+// For app pod veth pair(in host ns) ingress hook
+SEC("tc")
+int app_outbound(struct __sk_buff *skb)
+{
+    void *data = (void *)(long)skb->data;
+    void *data_end = (void *)(long)skb->data_end;
+    struct ethhdr *eth = data;
+    struct iphdr *iph;
+    __u32 outbound_cb = OUTBOUND_CB;
+    struct ztunnel_info *zi = get_ztunnel_info();
+
+    if (!zi || zi->ifindex == 0)
+        return TC_ACT_OK;
+
+    if (data + sizeof(*eth) > data_end)
+        return TC_ACT_OK;
+
+    // TODO: support ipv6
+    if (eth->h_proto != bpf_htons(ETH_P_IP))
+        return TC_ACT_OK;
+
+    iph = data + sizeof(*eth);
+    if (data + sizeof(*eth) + sizeof(*iph) > data_end)
+        return TC_ACT_OK;
+
+    if (iph->protocol != IPPROTO_TCP)
+        return TC_ACT_OK;
+
+    __builtin_memcpy(eth->h_dest, zi->mac_addr, ETH_ALEN);
+
+    skb->cb[4] = outbound_cb;
+
+    dbg("outbound redirect to ztunnel ifidx: %u\n", zi->ifindex);
+    return bpf_redirect(zi->ifindex, 0);
+}
+
+// For app pod veth pair(in host ns) egress hook
+SEC("tc")
+int app_inbound(struct __sk_buff *skb)
+{
+    void *data = (void *)(long)skb->data;
+    void *data_end = (void *)(long)skb->data_end;
+    struct ethhdr  *eth = data;
+    struct iphdr *iph = NULL;
+    struct ztunnel_info *zi = NULL;
+    __u32 inbound_cb = INBOUND_CB;
+
+    if (skb->mark == ZTUNNEL_BYPASS_MARK)
```

Comment:
Not a huge deal, but I'm wondering if it makes sense to use cb here instead of mark since we're already using cb for inbound vs outbound?  Is there any concern that mark will interfere with things going on within the pod netns?

Reply:
Definitely, using cb makes sense as a unified style to classify/mark pkts between app and ztunnel.